File: src/main/java/Solution_234_PalindromeLinkedList.java
public class Solution_234_PalindromeLinkedList {
    public static void main(String[] args) {
        var node1 = new ListNode2(1);
        var node2 = new ListNode2(2);
        var node3 = new ListNode2(3);
        var node4 = new ListNode2(4);
        var node5 = new ListNode2(5);
        node1.setNext(node2);
        node2.setNext(node3);
        node3.setNext(node4);
        node4.setNext(node5);
        System.out.println(isPalindrome(node1));
    }

    private static boolean isPalindrome(ListNode2 head) {
        if (head.next == null) return true;
        var size = 0;
        var current = head;

        while (current != null) {
            size++;
            current = current.next;
        }

        var nums = new int[size];
        current = head;

        for (int i = 0; i < size; i++) {
            nums[i] = current.val;
            current = current.next;
        }

        for (int i = 0; i <= size / 2; i++) {
            if (nums[i] != nums[size - i - 1]) return false;
        }

        return true;
    }
}

File: src/main/kotlin/Solution_234_PalindromeLinkedList.kt
/**
 * 234 Palindrome Linked List
 *
 * Easy
 *
 * Given the head of a singly linked list, return true if it is a palindrome or false otherwise.
 *
 *
 *
 * Example 1:
 *
 *
 * Input: head = [1,2,2,1]
 * Output: true
 * Example 2:
 *
 *
 * Input: head = [1,2]
 * Output: false
 *
 *
 * Constraints:
 *
 * The number of nodes in the list is in the range [1, 10^5].
 * 0 <= Node.val <= 9
 *
 *
 * Follow up: Could you do it in O(n) time and O(1) space?
 */
fun main() {
    val node1 = ListNode(1)
    val node2 = ListNode(1)
    val node3 = ListNode(0)
    val node4 = ListNode(1)
//    val node5 = ListNode(2)
//    val node6 = ListNode(1)
    node1.next = node2
    node2.next = node3
    node3.next = node4
//    node4.next = node5
//    node5.next = node6
    println(isPalindrome(node1))
}

fun isPalindrome(head: ListNode?): Boolean {
    if (head?.next == null) return true
    var current = head
    var size = 1

    // считаем размер списка
    while (current?.next != null) {
        size++
        current = current.next
    }

    // добавляем значения из списка в массив
    val digits = IntArray(size)
    current = head
    for (i in 0 until size) {
        digits[i] = current!!.`val`
        current = current.next
    }

    // сравниваем крайние значения в массиве
    for (i in 0 until size) {
        if (digits[i] != digits[size - i - 1]) return false
    }

    return true
}